src/shader_recompiler/ir/passes/resource_tracking_pass.cpp

```diff
@@ -129,32 +129,10 @@ bool IsImageInstruction(const IR::Inst& inst) {
 }
 
 IR::Value SwizzleVector(IR::IREmitter& ir, auto sharp, IR::Value texel) {
-    boost::container::static_vector<IR::Value, 4> comps;
-    for (u32 i = 0; i < 4; i++) {
-        switch (sharp.GetSwizzle(i)) {
-        case AmdGpu::CompSwizzle::Zero:
-            comps.emplace_back(ir.Imm32(0.f));
-            break;
-        case AmdGpu::CompSwizzle::One:
-            comps.emplace_back(ir.Imm32(1.f));
-            break;
-        case AmdGpu::CompSwizzle::Red:
-            comps.emplace_back(ir.CompositeExtract(texel, 0));
-            break;
-        case AmdGpu::CompSwizzle::Green:
-            comps.emplace_back(ir.CompositeExtract(texel, 1));
-            break;
-        case AmdGpu::CompSwizzle::Blue:
-            comps.emplace_back(ir.CompositeExtract(texel, 2));
-            break;
-        case AmdGpu::CompSwizzle::Alpha:
-            comps.emplace_back(ir.CompositeExtract(texel, 3));
-            break;
-        default:
-            UNREACHABLE();
-        }
-    }
-    return ir.CompositeConstruct(comps[0], comps[1], comps[2], comps[3]);
+    const std::array components = sharp.DstSelect().template Apply<IR::Value>(
```

Comment:
For swapping an IR vector I think I have another idea. You can introduce a new OpCompositeShuffle instruction, in a similar way to how OpCompositeExtract class of functions are defined, that works the same as OpVectorShuffle. Zero/One have indices of 0, 1 while Red and up have 4+ so the first vector can be an auxiliary vector of constants and put the value value into second slot, so will both reduce number of spirv instructions emitted and reduce needed code here